Added logging + config file + first tests

sbellone opened the pull request:
Moved manual testing from mockactors classes to test/manual
Using 'winston' for logging
Using 'config' for configuration
Firsts 'mocha' tests

Commit:
Added logging + config file + first tests

ovaussy-virtuoz commented on config/index.js:
```diff
@@ -8,16 +8,17 @@
 {
-  port: 8080,
-  amqp: { url: 'amqp://10.50.61.102'},
+  port: CONFIG.io.port,
```
Here we want to take the port define as an environment variable is defined. look at what we do for kqueryServer if you want. This is nodejs standart stuff so that people can change easily the port even with exactly the same configuration

ovaussy-virtuoz commented on config/index.js:
```diff
@@ -8,16 +8,17 @@
-    { name: 'agent', ids: [ 456 ], queue: QUEUEFLAG.actor | QUEUEFLAG.endpoint },
-    { name: 'livechat', ids: [ 456 ], queue: QUEUEFLAG.endpoint },
-    { name: 'ui', ids: [ 456 ] }
+    { name: CONFIG.endpoints.agent.name, ids: [ CONFIG.endpoints.agent.id, 457 ], queue: QUEUEFLAG.actor | QUEUEFLAG.endpoint },
```
shouldn't thos be completely configurable ?

ovaussy-virtuoz replied:
I mean in all those endpoint in a config files?

callixte replied:
Yes it would be better to pass all the endpoints at once as an array.

sbellone replied:
Sure, I just quickly commited what I had before leaving.

ovaussy-virtuoz commented on mockactors/livechat.js:
```diff
@@ -37,5 +37,4 @@ var LiveChat = new JS.Class(Actor, {
   }
```
What's the meaning of mockator ?

callixte replied:
Mock actors

callixte replied:
The complete classes could have been moved to tests. Actor.js needs to be moved to lib/.

ovaussy-virtuoz commented on test/manual/ui_test.js:
```diff
@@ -0,0 +1,30 @@
+
+var UI = require('../../mockactors/ui').UI;
+
+var ui = new UI('localhost:8080', 'ui/456', '10.50.61.103');
```
Can you add some comment to explain what you're testing manually here?

ovaussy-virtuoz commented on test/manual/wbh2_test.js:
```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright 2012 VirtuOz Inc. All rights reserved.
```
Same , please add some comment

ovaussy-virtuoz commented on test/smartrouter_test.js:
```diff
@@ -0,0 +1,109 @@
+      done();
+    });
+
+    smartrouter.start(config);
```
Mmmm where are your asserts here ?

ovaussy-virtuoz replied:
Don't you want to check that the logger was correctly called ? if no then forget my comment.

ovaussy-virtuoz commented on util/logger.js:
```diff
@@ -0,0 +1,61 @@
+        {
+          level: CONFIG.level,
+          filename: CONFIG.filename,
+          json: false
```
why false ? I like json

callixte commented on lib/smartrouter.js:
```diff
@@ -28,17 +30,32 @@ var SmartRouter = new JS.Class(EventEmitter, {
+
+  configureIo: function() {
+    var self = this;
+    self.io.configure(function() {
```
See https://github.com/LearnBoost/Socket.IO/wiki/Configuring-Socket.IO for how to configure socket.io.

sbellone replied:
That's where I looked. I will add the different configurations 'development' and 'production'

callixte commented on config/development.yaml:
```diff
@@ -0,0 +1,23 @@
```
Your refactor on how to configure does not go in the right direction.
The idea of the configuration is that the Samrt Router as-is is dumb and can handle any business logic. It can be open-sourced and used for any purpose. It is our configuration that will then describe our business logic with agents, uis, livechat providers etc.
The idea was that the final user has to implement a way to fill this config object and pass it to the smartrouter. The way can be yaml, json or retrieving it from a store (which is our target usage).
Let it like that for now. I'll give it a crack later.

callixte commented on test/smartrouter_test.js:
```diff
@@ -0,0 +1,109 @@
+var mockedUI;
+var mockedLiveChat;
+
+// /!\ Those tests need to have a local RabbitMQ running
```
I do not agree with this, especially knowing how heavy weight is Rabbit. Plus, it would have to be different on three different OSes we use to develop these node projects.

callixte replied:
I challenge you to do so in a manner that supports Windows, MacOSX and Linux. Extra credit if Erlang is not required prior hand.

sbellone replied:
Actually the problem is to have Erlang installed. It's heavy and cannot be packaged.

But as the user will need to have RabbitMQ installed and running to run the Smartrouter anyway, I don't see why it is a problem to ask him to have it installed before launching the tests.
We could do a download/unzip/run script, but why not test with the version the user will use in "production"? Just have a rabbitmq_path property and run it in the before() method seems fine to me.

sbellone commented on test/manual/ui_test.js:
```diff
@@ -0,0 +1,30 @@
+  ui.talk('woohoo');
+}, 10000);
+
+/*setInterval(function () {
```
It was in the prototype. It's an example of what we can do I guess.

sbellone commented on test/smartrouter_test.js:
```diff
@@ -0,0 +1,109 @@
+  it('should connect an agent to the smartrouter', function(done)
+  {
+    console.log('starting agent test');
+    mockedAgent = new Agent('localhost:8080', 'agent/456', '10.50.61.103');
```
It's the id of the Actor. Callixte put an IP but it can be anything actually.

sbellone commented on test/smartrouter_test.js:
```diff
@@ -0,0 +1,109 @@
+    mockedUI.talk('Hey is there someone?');
+  });
+
+//  it('should send a message from the agent to the UI', function(done)
```
Does not work yet because I did not succeed to shutdown/restart the smartrouter (with beforeEach and afterEach methods). Will do that today.

sbellone commented on util/logger.js:
```diff
@@ -0,0 +1,61 @@
+ * @created 2012-11-08
+ */
+
+var winston = require('winston'),
```
I was wondering indeed. But I am waiting to see if it will be the only dependency on vnodelib. If yes, maybe it doesn't worth making a dependency on the whole vnodelib just for this class.

Commit:
Added tests. Moved test mocks

Commit:
So I have changed a few things around in your stop() method and it now work.
Sort of. The problem we are hitting is that the connections are pooled in the same node process. So everytime you use localhost:8080, it uses the same websocket, even if what you see at the socket.io is different. It is why it is now working with two tests as I specified localhost and 127.0.0.1 which are different. There may be a config option not to pool.

Commit:
With the param to enforce new connections

But now the tests are failing: you need to instantiate new clients in each test.

Commit:
Tests are passing. Moved socket.io testing parameters into the test class.

Commit:
Method setup() of actor.js renamed to connect() and io.connect() now called in it.

Commit:
Started to add better logging

Commit:
Reverted endpoints config, improved logging, started detailed documentation.

Commit:
Tests of shutdown/restart without losing data + documentation + very basic example with 2 actors

Commit:
- Added empty object when emitting events 'whoareyou' and 'hello' because
  SocketIO4Net don't receive events without payload.
- Added new method to be able to register new endpoint's ids without
  restarting the smart-router

Commit:
fixing a glitch in case connectionParams are not passed on Actor creation.
